src/adaptation/icap/OptXact.cc

```diff
@@ -105,17 +106,36 @@ bool Adaptation::Icap::OptXact::parseResponse()
     debugs(93, 5, "have " << readBuf.length() << " bytes to parse" << status());
     debugs(93, DBG_DATA, "\n" << readBuf);
 
+    if (!replyParser)
+        replyParser = new ResponseParser();
+
+    Must(replyParser->parse(readBuf)); // throws on parse failure OR incomplete data
+    readBuf = replyParser->remaining(); // leave any unparsed bytes in the I/O buffer
+
+    /* We know the whole response is in parser now */
+    // XXX: server connection details not available in this scope
+    // debugs(93, 2, "ICAP Server " << connection);
+    debugs(93, 2, "ICAP Server RESPONSE:\n---------\n" <<
+           replyParser->messageProtocol() << " " << replyParser->messageStatus() << " " << replyParser->reasonPhrase() << "\n" <<
+           replyParser->mimeHeader() <<
+           "----------");
+
     HttpReply::Pointer r(new HttpReply);
     r->protoPrefix = "ICAP/"; // TODO: make an IcapReply class?
+    r->sline.set(replyParser->messageProtocol(), replyParser->messageStatus());
 
-    if (!parseHttpMsg(r.getRaw())) // throws on errors
-        return false;
+    // parse headers
+    if (!r->parseHeader(*replyParser)) {
+        r->sline.set(replyParser->messageProtocol(), Http::scInvalidHeader);
+        debugs(93, 2, "error parsing response headers mime block");
+    }
```

Comment:
I am resurrecting/continuing an out-of-date [change request](https://github.com/squid-cache/squid/pull/682/#discussion_r585753086) here to ease navigation. I resolved that change request.

> Alex: AFAICT, this replacement of the parseHttpMsg() call is incorrect; it does not follow parseHttpMsg() API/invariants. I do not know whether we should keep parseHttpMsg() or not, but I do not want to write parsing code using change requests. Have you tried preserving parseHttpMsg() calls?

> Amos: AFAICT the parseHttpMsg() attempted to parse the buffer and chopped out the bytes which appeared to be a valid response (status-line + headers block). That is all performed by the new liens 112-113 above

```C++
Must(replyParser->parse(readBuf)); // throws on parse failure OR incomplete data
readBuf = replyParser->remaining(); // leave any unparsed bytes in the I/O buffer
```

No, those lines do not seem to follow parseHttpMsg() API/invariants. For example, parseHttpMsg() has three outcomes (need-more-data, success, exception). The lines you mention only have two outcomes (success and exception). The surrounding code is unaware of this major change AFAICT, hence, is likely broken. BTW, the difference in outcomes explains why the Adaptation::Icap::OptXact::parseResponse() always returns true now, highlighting this part of the problem.

There are other potential parseHttpMsg() API/invariant changes that I have not investigated yet.

I am sure this all can be fixed, one way or another, but, again, I would very much prefer not to progress bug-by-bug here: It is very difficult and painful to write correct parsers via code review, for both the author and the reviewer. If you can largely preserve parseHttpMsg() calls, it may help a lot. The primary goals of this PR can be satisfied without significant changes to parseHttpMsg() API/invariants AFAICT. And if they cannot be, then we may have to split this work in two phases (more pain), to carefully control the switch from parseHttpMsg() to something better.